rfcs/proposed/task_group_dynamic_dependencies/task_handle_semantics.md

```diff
@@ -0,0 +1,445 @@
+
+As it was stated in the parent RFC document, we would like to allow adding predecessors in any state described above and to limit the successor to be a task in created state since it can be too late to add predecessors to
+the task in executing or completed state.
+
+The second limitation is handled by limiting the successor argument to be only ``task_handle``.
+
+Lets consider the different states of the task tracked or handled by ``predecessor``. 
+
+If the predecessor task is in any state except the completed one (created/scheduled/running), the API registers the successor task
+in the list of successors on the predecessor side and increase the corresponding reference counter on the successor side to ensure it
+would not be executed before the predecessor task. The successor task can only start executing once the associated reference counter is equal to 0.
+
+If the predecessor task is in `completed` state, the API has no effect in terms of modifying the list of successors and reference counters since no additional
+dependencies required and the successor task can be executed if all other dependent tasks are executed as well.
+
+If the predecessor task state has changed while registering the task as a predecessor for any task, the API should react accordingly to make sure
+adding dependencies and increasing the corresponding reference counters are not done for completed tasks.
+
+Implementation-wise, this API requires adding a list of successors into the predecessor task and adding the new vertex instance that corresponds
+to the successor task. This vertex would contain the reference counter and a pointer to the successor task itself. Each element in the task successor list
+is a pointer to the vertex instance of the successor task.
+
+The vertex instance is created once the first task is registered as a predecessor and is reused by any other predecessors. 
+
+Once the predecessor task is completed, it should go through the list of successor vertices and decrement the reference counter. Once the successor's
+reference counter is equal to 0, the successor task can be scheduled for execution.
+
+API-wise, the function that decreases the reference counter may also return the pointer to the task. If the reference counter is not equal to 0, the
+returned pointer is ``nullptr``. Otherwise, the successor task pointer is returned. It is required to allow bypassing one of the successor tasks
+if the body of the predecessor task did not return other task that should be bypassed.
+
+This implementation approach is illustrated in the picture below:
+
+<img src="predecessor_successor_implementation.png" width=800>
+
+### Adding successors to the current task
+
+Consider use-case of parallel wavefront pattern on the 2-d grid. Each cell is computed as part of a separate task in ``task_group``. Each cell task computes
+itself and creates more tasks to process the cell below and the cell on the right.
```

Comment:
Usually the executing task does not create the other tasks in a wavefront, since there are typically multiple predecessors. For example, the task to the right of a task A, let's call it C, may also be the task below B. More typically, wavefront tasks are pre-allocated and dependencies set from outside of the tasks themselves. Otherwise, who is responsible for creating C? A or B?  What if A creates C, adds itself as a predecessor, and completes before B executes? Would there be an edge from B to C?

Reply:
We have found an example of a divide-and-conquer wavefront where edges are made from within an executing task and also from external tasks, so my previous comment is moot.

Reply:
The `add_successor` API was removed from the proposal. I have added the separate section describing alternatives.
I have also added the "Advanced examples" section describing all wavefront patterns we found with the possible implementations.